### PR TITLE
New workflow implementation and migration fixes

### LINKED ISSUE #4226: MariaDB upgrade script for 12.x fails with "Data too long for column 'if_condition'"
**Describe the bug**
When running the python update script "_workflows_mariadb.py_" for converting database from opencast 11 to 12 I encounter the following error:

```
Migrating workflow 59187586 (14837/14849)
Migrating workflow 59187818 (14838/14849)
Migrating workflow 59187821 (14839/14849)
Traceback (most recent call last):
File "workflows_mariadb.py", line 368, in <module>
insert_parsed(create_workflow_operation_sql, workflow_operations)
File "workflows_mariadb.py", line 67, in insert_parsed
cursor.executemany(sql, data)
File "/usr/local/lib/python3.6/site-packages/mysql/connector/cursor.py", line 674, in executemany
return self.execute(stmt)
File "/usr/local/lib/python3.6/site-packages/mysql/connector/cursor.py", line 572, in execute
self._handle_result(self._connection.cmd_query(stmt))
File "/usr/local/lib/python3.6/site-packages/mysql/connector/connection.py", line 922, in cmd_query
result = self._handle_result(self._send_cmd(ServerCmd.QUERY, query))
File "/usr/local/lib/python3.6/site-packages/mysql/connector/connection.py", line 732, in _handle_result
raise errors.get_exception(packet)
mysql.connector.errors.DataError: 1406 (22001): Data too long for column 'if_condition' at row 16
```

After dropping this workflow number 59187821 from "oc_jobs" table the migration runs without an issue.

**Environment Information**
 - OS: Alma Linux 8.6
 - Opencast 12.2.0 - a4c7257 - from RPM Repo
 - Python 3.6.8
 - MariaDB Server version: 10.3.32-MariaDB-log
 - mysql_connector_python==8.0.29
 - update script downloaded directly from github

### PR DESCRIPTION
**Pull request by mtneug**
This is based on #4179 as both add DB migrations

This PR fixes a bunch of issues with the new workflow implementation / the migration script:

* Allow setting a port number in the migration script (CockroachDB does not use the default PostgreSQL port)
* Change data types of some columns in the workflow tables to allow for workflow definitions using longer values
* Add and normalize DB indexes in MariaDB and PostgreSQL
* Fix exception due to null values in configuration tables (see commit message for exception log)

This PR includes migrations for 12.2 to hopefully 12.3 (optional) as well as 12 to 13. The `RENAME INDEX` queries for MariaDB are only available since 10.5.2 (released 2020-03-26). According to the documentation, there is also no `IF EXISTS` which will probably lead to a failed execution if both the 12.2 -> 12.3 and the 12 -> 13 migrations are executed. Do you know of a better way?

Fixes #4226

### Your pull request should…

* [ ] have a concise title
* [ ] [close an accompanying issue](https://help.github.com/en/articles/closing-issues-using-keywords) if one exists
* [ ] [be against the correct branch](https://docs.opencast.org/develop/developer/development-process#acceptance-criteria-for-patches-in-different-versions)
* [ ] include migration scripts and documentation, if appropriate
* [ ] pass automated tests
* [ ] have a clean commit history
* [ ] [have proper commit messages (title and body) for all commits](https://medium.com/@steveamaza/e028865e5791)

### CODE REVIEW

**Review by lkiesow — CHANGES_REQUESTED** [review]
- I didn't test this, but I'm pretty sure your manual modifications do not fit the auto-generated table structure any longer
  - You also vary types between PostgreSQL and MariaDB
- Can we have EclipseLink automatically make these adjustments?
- What about documenting these as optional for 12 in the release notes (but adding them to migration and auto-generation) and having them as default scripts for 13?

**Comment by mtneug** [comment]
> I didn't test this, but I'm pretty sure your manual modifications do not fit the auto-generated table structure any longer

Will take a look.

> You also vary types between PostgreSQL and MariaDB

@lkiesow, what variation do you mean? PostgreSQL only has `TEXT`.

> Can we have EclipseLink automatically make these adjustments?

Not sure. At least when I tested creating indexes in #4179 this was not done automatically, but apparently this is a setting?

> What about documenting these as optional for 12 in the release notes (but adding them to migration and auto-generation) and having them as default scripts for 13?

+1 on adding this to the release notes. I still like to have the actually SQL migration files in this directory though.

File: docs/upgrade/11_to_12/workflows_mariadb.py [hunk 1]
```diff
@@ -164,9 +166,9 @@ def parse_bool(value):
   `continuable` tinyint(1) DEFAULT 0,
   `completed` datetime DEFAULT NULL,
   `started` datetime DEFAULT NULL,
-  `description` varchar(255) DEFAULT NULL,
+  `description` longtext DEFAULT NULL,
   `exception_handler_workflow` varchar(255) DEFAULT NULL,
-  `if_condition` varchar(255) DEFAULT NULL,
+  `if_condition` varchar(65535) DEFAULT NULL,
```
Comment (Rillke) [inline docs/upgrade/11_to_12/workflows_mariadb.py:171]: This caused:

```
Traceback (most recent call last):
  File "/usr/local/lib/python3.8/dist-packages/mysql/connector/connection_cext.py", line 535, in cmd_query
    self._cmysql.query(query,
_mysql_connector.MySQLInterfaceError: Column length too big for column 'if_condition' (max = 16383); use BLOB or TEXT instead
```

After setting `varchar(16383)` I got

```
Traceback (most recent call last):
  File "/usr/local/lib/python3.8/dist-packages/mysql/connector/connection_cext.py", line 535, in cmd_query
    self._cmysql.query(query,
_mysql_connector.MySQLInterfaceError: Row size too large. The maximum row size for the used table type, not counting BLOBs, is 65535. This includes storage overhead, check the manual. You have to change some columns to TEXT or BLOBs
```

so I used `varchar(2555)`

File: docs/upgrade/12.2_to_12.3_optional/mariadb.sql [hunk 2]
```diff
@@ -0,0 +1,27 @@
+CREATE INDEX IF NOT EXISTS IX_oc_job_argument_id ON oc_job_argument (id);
+CREATE INDEX IF NOT EXISTS IX_oc_job_argument_id_argument ON oc_job_argument (id, argument);
```
Comment (Rillke) [inline docs/upgrade/12.2_to_12.3_optional/mariadb.sql:2]:
```
MariaDB [opencast]> CREATE INDEX IF NOT EXISTS IX_oc_job_argument_id_argument ON oc_job_argument (id, argument);
ERROR 1071 (42000): Specified key was too long; max key length is 3072 bytes
```

Comment (geichelberger) [reply docs/upgrade/12.2_to_12.3_optional/mariadb.sql:2]: With Postgres it's the same:
opencast=# CREATE INDEX IF NOT EXISTS IX_oc_job_argument_id_argument ON oc_job_argument (id, argument);
ERROR: Index row requires 19768 bytes maximum size is 8191

**Comment by mtneug** [comment]
Closed in favor or #4456.

**Pull request closed without merging**